Improved, pipelined AAE fullsync

jtuple opened the pull request:
This pull-request includes multiple improves to AAE fullsync, including:
1. Making AAE exchange pipelined. This is the most significant change, since non-pipelined AAE is sensitive to network latency. And in the worst case was shown to take weeks to complete, even for a trivial exchange. The new pipelined approach is designed to handle high latency links.
2. Ensuring AAE fullsync only does at most 1 bloom fold per partition. Previously, the implementation did a bloom fold per AAE tree per partition.
3. Directly sending up to a configurable threshold of key differences, thus avoiding the bloom fold entirely when clusters are mostly in sync.
4. Using the internal `n_val=1` get option when retrieving objects to send. This sends only a single request to a single vnode, therefore reducing cluster load during a fullsync.

Preliminary test result can be found [here](https://gist.github.com/jtuple/92e0fe300579ed24ef80). More testing is underway/planned.

Sibling pull-requests: basho/riak_core#628, basho/riak_kv#1023

jtuple committed:
Extend AAE fullsync to use a pipelined exchange

The current AAE fullsync exchange was based on the exchange FSM used by
Riak for local cluster AAE. However, this was a bad design decision as
local AAE was never designed to tolerate high latency links.

In the worst case, this design can require millions of sequential
roundtrip messages. This is not an issue for a fast LAN, but is
impractical for WANs with 10-200ms RTTs.

This commit changes the AAE fullsync exchange to use a pipelined approach.
Rather than performing a synchronous request for each bucket, the exchange
determines all differences for a given level of the AAE tree and then
requests the data for all differing buckets upfront.

Ideally, this would supported by a new streaming protocol. However to
retain protocol compatibility the pipelined approach sends all the
known bucket requests upfront and relies upon TCP ordering to deliver
the responses in the order requested.

Note: this commit relies upon changes made to riak_core and riak_kv to
extend the AAE subsystem to support the new level-by-level exchange that
is utilized by this new fullsync approach.

jtuple committed:
Refactor AAE FS source state transitions for clarity

This commit slightly refactors the riak_repl_aae_source code to
make state transitions more obvious.

This change also eases the porting of commits from a separate feature
branch that implemented a slightly different approach to improved AAE
syncing.

jtuple committed:
Add buffer-then-send mode for AAE fullsync repair

This commit merges in the separately prototyped buffer-then-send
approach to directly sending key differences. The mode used is
determined by the `fullsync_direct_mode` application variable,
which defaults to `inline`.

* inline
  differences selected for direct repair are sent as they are encountered
  during an AAE exchange.

* buffered
  differences selected for direct repair are buffered in memory during an
  exchange. After the exchange, the buffer is sorted and keys are then sent
  in order.

krestenkrab committed:
Only do one backend fold for each vnode, not N

Previous code created a bloom filter for each
IndexN, populating it with diffs, and then did
a backend fold for that vnode sending objects
along the way.

The code now

1. creates the bloom filter in init()
2. runs key exchange for each IndexN
3. then does a vnode fold to find objects in the
   bloom filter.

krestenkrab committed:
Use N=1 for reading values for replication

krestenkrab committed:
Adaptable sync strategy: get or fold w/threshold

This change introduces a threshold for how
to process fullsync diffs.  A small number of
changed KVs (currently fixed at 1000 per vnode)
are read using GET; above that limit, objects
are read using a backend fold.

For small number og changed KVs, this avoids
doing a full scan of the backend data.

jtuple committed:
Make fullsync direct send threshold configurable

jtuple committed:
Fix deadlock between AAE fullsync sink/source

A deadlock in AAE fullsync was identified during testing.

The deadlock occurs when both the AAE fullsync source and sink
processes block on TCP send due to TCP backpressure. Since both
processes are blocked, neither can perform a receive to unblock
the other.

This commit addresses the issue by making the AAE sink use a helper
process for sending, therefore ensuring the sink never blocks on a
send. The sink will therefore always get around to receiving data at
some point, unblocking the source if needed.

NOTE: the SSL transport wraps the underlying socket in a gen_server, and
all socket functions route through this server. As such, the SSL transport
is still vulnerable to this deadlock. Fixing this requires either changing
the AAE protocol or patching the OTP SSL implementation. One of the two is
planned future work.

jtuple committed:
Remove non-pipelined AAE fullsync mode

krestenkrab commented:
Maybe we should use `riak_kv_vnode:local_get(PartitionIndex,BKey)` rather than `Client:get(..., [{n_val,1}...`.  This would even further reduce the cluster load, as we don't risk going over the network to get the value.  We are, after all, running in the local vnode that is supposed to have the value.  The only obvious downside I can see is that `local_get`doesn't let you specify a timeout.

krestenkrab commented:
I'd like to propose that we rename `fullsync_direct` to be `fullsync_direct_limit`.  In our next round of changes we have a sibling to this config option which is the cutoff percentage (of #keys in partition); and that one I'd like to call `fullsync_direct_percentage_limit`.

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -30,6 +30,15 @@
 -type index_n() :: {index(), pos_integer()}.
 
+-record(exchange, {mode   :: inline | buffered,
+                   buffer :: ets:tid(),
```
buffer can also be undefined.

jonmeredith commented on src/riak_repl_aae_sink.erl:
```diff
@@ -164,8 +167,41 @@ process_msg(?MSG_COMPLETE, State=#state{owner=Owner}) ->
+    spawn_link(?MODULE, sender_init, [Transport, Socket]).
+
+sender_init(Transport, Socket) ->
+    sender_loop({Transport, Socket}).
```
Be kind to your fellow debuggers and set something in the process dictionary we can postmortem what this process is and why it existed.

krestenkrab replied:
I suppose this means: use `proc_lib`.

jonmeredith replied:
That's what I used to do until the inestimable @Vagabond pointed out that anything started with `proclib` should comply with OTP Design principles and handle ``sys` messages etc.
Maybe a poor mans version would set the the procdict entries so that etop etc showed them nicely.

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -305,29 +353,60 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
+    send_asynchronous_msg(?MSG_GET_AAE_BUCKET, {Level,Bucket,IndexN}, State).
+
+wait_get_bucket(_Level, _Bucket, _IndexN, State) ->
+    Reply = get_reply(State),
```
Any need for Reply variable?

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -305,29 +353,60 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
+    send_asynchronous_msg(?MSG_GET_AAE_SEGMENT, {Segment,IndexN}, State).
+
+wait_get_segment(_Segment, _IndexN, State) ->
+    Reply = get_reply(State),
```
unneeded var

krestenkrab commented on src/riak_repl_aae_source.erl:
```diff
@@ -494,7 +597,7 @@ send_missing(RObj, State=#state{client=Client, wire_ver=Ver, proto=Proto}) ->
 send_missing(Bucket, Key, State=#state{client=Client, wire_ver=Ver, proto=Proto}) ->
-    case Client:get(Bucket, Key, 1, ?REPL_FSM_TIMEOUT) of
+    case Client:get(Bucket, Key, [{r, 1}, {timeout, ?REPL_FSM_TIMEOUT}, {n_val, 1}]) of
         {ok, RObj} ->
```
I was reading around some old documentation for full sync, and came across the `max_get_workers` full sync configuration property.  This seems to have fallen through the cracks in the `v3` full sync scheme; perhaps this may also be a bottle neck for performance.

jonmeredith replied:
I think that setting controlled how many workers the fullsync source used for retrieving vclocks and objects to send.

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -443,36 +526,56 @@ replicate_diff(KeyDiff, Acc, State=#state{index=Partition}) ->
+            Exchange2 = handle_direct(Mode, Bucket, Key, Exchange, State),
+            Exchange2#exchange{count=Count+1};
+       true ->
+            %% Past threshold, add to bloom filter for future bloom fold
```
Could build the bloom filter here on demand instead of in `maybe_accumulate_key/4` - would save allocating around 100Mb on top of the potential stack used by compare2 of 16mb.

```
(riak@127.0.0.1)84>     NumKeys = 10000000.
(riak@127.0.0.1)84>     {ok, Bloom} = ebloom:new(NumKeys, 0.01, random:uniform(1000)).
(riak@127.0.0.1)86> ebloom:size(Bloom).
95929552
```

Would be better for the frequent-small-syncs case.

krestenkrab replied:
Right. This is also part of the 'estimate keys' extension to this.  Makes good sense to add that optimisation here as well, now that we have the 'fullsync_direct_limit' config option.

The 1M fixed limit is pretty bad for our primary use case w/ 2B keys, N=3, Q=128. That leaves ~50M keys in each partition; meaning that a 1M bloom has 50x chance of a false positive, i.e. a false positive probability of 0.5.  Fortunately, the actual number of diffs we'll see is very small, as we also run realtime sync.

jonmeredith commented:
Couldn't work out how to inline respond to Kresten's comment, but yes, I think it does need to use `riak_kv_vnode:local_get/2` as there are no guarantees you will hit the local vnode if you use a reduced n_val get (which is probably something we should also fix for CS).

jonmeredith commented:
Unrelated to the AAE changes, the mechanics for changing the controlling process are totally unnecessary for the `Remote` function defined in `key_exchange/2` as there are no messages delivered by an `{active, false}` socket.

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -305,29 +353,60 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
+    lager:info("Directly sent ~b differences inline for partition ~p",
+               [Sent, Partition]),
+    ok;
+maybe_send_direct(#exchange{buffer=Buffer}, State=#state{index=Partition}) ->
```
If you get here and you have a bloom filter, you do the direct gets (at the moment using get FSMs) for all of the entries.  Is there a good reason not to detect you will need to bloom fold anyway and just set the elements?

jonmeredith commented on src/riak_repl_aae_source.erl:
```diff
@@ -305,29 +353,60 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
+                  #state{index=Partition}) ->
+    %% Inline sends already occured inline
+    Sent = erlang:min(Count, Limit),
+    lager:info("Directly sent ~b differences inline for partition ~p",
```
Having multiple sent counts per partition is confusing for users and monitoring.  You will potentially have log messages that say

`Directly sent 1000 differences inline for partition 123456789
No differences, skipping bloom fold for partition 123456789`

`Directly sending 1000 differences inline for partition 123456789
No differences, skipping bloom fold for partition 123456789`

`Directly sent 1000 differences inline for partition 123456789
Bloom folding over ~54321 differences for partition 123456789`

`Directly sending 1000 differences inline for partition 123456789
Bloom folding over ~54321 differences for partition 123456789`

So… as a user, I would like just one log message that succinctly tells me what you completed sending (not intended sending) and if you like, you can add some details about the way you did it.

```
Fullsync sent 55321 differences for partition 123456789 in 13.4 seconds using [direct,bloom]
```

krestenkrab replied:
Rune and I also talked about reporting. It would make sense to report something that can help guide selecting configuration values. In particular direct/bloom cutoff limit, because if often overshoot by a small number - say the average # diffs is 1200 and the limit is configured to 1000 then it makes sense to set the cutoff at ~1400 perhaps if that means that you avoid a huge fold in most cases.  

One heuristic to apply would be to - if folding - count the actual number of objects folded over, and if the ratio of diffs/fold-count is say below 1% issue a warning that it may make sense to raise the full sync_direct_limit.

jonmeredith commented:
Spoke with @jtuple and @matthewvon on the path to merging.  This is relatively minimal change PR to a point release to address a significant pain point, so I'm accepting that we won't clean up some things I would like to.

Code review looks logically good.  The only major thing that needs to be fixed is the switch to `riak_kv_vnode:local_get/2`.  The reporting improvements and general refactoring can be done on the 2.1 schedule.  Depending on how much improvement is needed to the protocol the module may be significantly refactored anyway.  Joe said the AAE estimation branch moved the bloom filter creation, so if we get that merged it will take care of that.

Will approve once I've run some of the r_t AAE tests.

jtuple committed:
Change AAE FS source to use direct vnode gets

When performing direct sends, the AAE fullsync source retrieves
individual objects in order to send them. Previously, this was
accomplished using normal riak_client gets using the n_val=1
option. However, there are no guarantees about which replica
the object would come from using such an approach.

This commit changes the AAE fullsync source to instead retrieve
objects directly from the vnode performing the fullsync. This
ensures that the object sent aligns with the AAE hash used to
determine differences. This approach may also have better
performance.

jonmeredith commented:
Compared riak_test output to 2.0.0 and get similar results for the intermittent cases. A little work to do there, but ok.

Like the other pull requests, this is targeted at develop, not 2.0.0

Plus one to merge once target is fixed.

engelsanchez commented:
I believe this is not needed anymore. All changes made it to 2.0, and today 2.0 has been merged to develop.
